Orange/widgets/utils/listfilter.py

```diff
@@ -0,0 +1,244 @@
+                for s1, s2 in reversed(list(slices(rows))):
+                    delslice(self.model(), s1, s2)
+            self.dragDropActionDidComplete.emit(res)
+
+    def dragEnterEvent(self, event):
+        """
+        Reimplemented from QListView.dragEnterEvent
+        """
+        if self.acceptsDropEvent(event):
+            event.accept()
+        else:
+            event.ignore()
+
+    def acceptsDropEvent(self, event):
+        """
+        Should the drop event be accepted?
+        """
+        # disallow drag/drops between windows
+        if event.source() is not None and \
+                event.source().window() is not self.window():
+            return False
+
+        mime = event.mimeData()
+        vars = mime.property('_items')
+        if vars is None:
+            return False
+
+        if not all(isinstance(var, self.__acceptedType) for var in vars):
+            return False
+
+        event.accept()
+        return True
+
+
+class VariableFilterProxyModel(QSortFilterProxyModel):
+    """ A proxy model for filtering a list of variables based on
+    their names and labels.
+
+    """
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        self._filter_string = ""
+
+    def set_filter_string(self, filter):
+        self._filter_string = str(filter).lower()
+        self.invalidateFilter()
+
+    def filter_accepts_variable(self, var):
+        row_str = var.name + " ".join(("%s=%s" % item)
+                                      for item in var.attributes.items())
+        row_str = row_str.lower()
+        filters = self._filter_string.split()
+
+        return all(f in row_str for f in filters)
+
+    def filterAcceptsRow(self, source_row, source_parent):
+        model = self.sourceModel()
+        if isinstance(model, VariableListModel):
+            var = model[source_row]
+            return self.filter_accepts_variable(var)
+        else:
+            return True
+
+
+class CompleterNavigator(QObject):
+    """ An event filter to be installed on a QLineEdit, to enable
+    Key up/ down to navigate between posible completions.
+    """
+    def eventFilter(self, obj, event):
+        if (event.type() == QEvent.KeyPress and
+                isinstance(obj, QLineEdit)):
+            if event.key() == Qt.Key_Down:
+                diff = 1
+            elif event.key() == Qt.Key_Up:
+                diff = -1
+            else:
+                return False
+            completer = obj.completer()
+            if completer is not None and completer.completionCount() > 0:
+                current = completer.currentRow()
+                current += diff
+                completer.setCurrentRow(current % completer.completionCount())
+                completer.complete()
+            return True
+        else:
+            return False
+
+
+class VariablesFilter(QObject):
+    def __init__(self, variable_list_model=VariableListModel(enable_dnd=True)):
```

Comment:
You should never use mutable values as default arguments (without having a very good reason for doing so, and being very careful).

What happens here is that *all* instances of `VariablesFilter` are gonna share the same instance of `VariableListModel`. Open canvas, connect two `SelectColumnsWidget` to `File` and have it visible at the same time. Move attributes around in one of them and observe how the other changes, too!